### ISSUE TITLE
Possible issue with accuracy computation

### ISSUE THREAD

**Issue by riccardorenzulli** [cywinski/SAeUron#5]
Hi,

While running **Step 5.1 – Hyperparameter sweep for object unlearning** (`run_acc_all_cls_sweep.py`), I noticed potential inconsistencies in how accuracy is computed.

Specifically, in [`accuracy_unlearncanvas_cls_sweep_fast.py#L120`](https://github.com/cywinski/SAeUron/blob/main/scripts/accuracy_unlearncanvas_cls_sweep_fast.py#L120), I sometimes get:
- **UA values < 0**
- **IRA values > 100%**

I think this comes from dividing by `len(classes_to_use)`.

### Suggested fix (Step 5.1)

```python
            for i in range(len(batch_class_labels)):
                object_class = class_available[batch_class_labels[i].item()]
                class_results["loss"][object_class] += class_loss[i].item()
                class_results["pred_loss"][object_class] += class_softmax[i][
                    batch_class_labels[i]
                ].item()
                # Start fix (Riccardo: I don't think dividing by classes_to_use 
                # is correct here, it should be divided by the number of used themes 
                # (limit-theme option), or as in my solution)
                # class_results["acc"][object_class] += (
                #     class_pred_success[i].item()
                #     * 1.0
                #     / (len(classes_to_use) * len(seed)) 
                # )
                class_results["acc"][object_class] += class_pred_success[i].item()
                # End fix (Riccardo)
                misclassified_as = class_available[class_pred_labels[i].item()]
                class_results["misclassified"][object_class][misclassified_as] += 1
    # Start fix (Riccardo)
    for object_class in class_available:
        total_samples_for_class = sum(class_results["misclassified"][object_class].values())
        
        if total_samples_for_class > 0:
            class_results["acc"][object_class] = class_results["acc"][object_class] / total_samples_for_class
    # End fix (Riccardo)
```

---

I also wonder if a similar adjustment is needed in
[`accuracy_unlearncanvas_cls_fast.py#L198`](https://github.com/cywinski/SAeUron/blob/main/scripts/accuracy_unlearncanvas_cls_fast.py#L198) (Step 6 – Run evaluation for object unlearning).

### Suggested fix (Step 6)

```python
    for batch_images, batch_style_labels in tqdm(style_dataloader):
        batch_images = batch_images.to(device)
        batch_style_labels = batch_style_labels.to(device)

        with torch.no_grad():
            # Style classification
            style_res = style_model(batch_images)
            style_loss = torch.nn.functional.cross_entropy(
                style_res, batch_style_labels, reduction="none"
            )
            style_softmax = torch.nn.functional.softmax(style_res, dim=1)
            style_pred_labels = torch.argmax(style_res, dim=1)
            style_pred_success = style_pred_labels == batch_style_labels
            
            for i in range(len(batch_style_labels)):
                # Update style results
                test_theme = theme_available[batch_style_labels[i].item()]
                style_results["loss"][test_theme] += style_loss[i].item()
                style_results["pred_loss"][test_theme] += style_softmax[i][
                    batch_style_labels[i]
                ].item()
                # Start fix (Riccardo: this should be ok, but I'm doing it my way)
                # style_results["acc"][test_theme] += (
                #     style_pred_success[i].item()
                #     * 1.0
                #     / (len(classes_to_use) * len(seed))
                # )
                style_results["acc"][test_theme] += style_pred_success[i].item()
                # End fix (Riccardo)
                misclassified_as = theme_available[style_pred_labels[i].item()]
                style_results["misclassified"][test_theme][misclassified_as] += 1
    # Start fix (Riccardo)
    for theme in theme_available:
        total_samples_for_theme = sum(style_results["misclassified"][theme].values())
        if total_samples_for_theme > 0:
            style_results["acc"][theme] = (style_results["acc"][theme] / total_samples_for_theme)
    # End fix (Riccardo)

    if not dry_run:
        style_output_path = os.path.join(output_dir, f"{cls}.pth")
        torch.save(style_results, style_output_path)

    for batch_images, batch_class_labels in tqdm(class_dataloader):
        batch_images = batch_images.to(device)
        batch_class_labels = batch_class_labels.to(device)

        with torch.no_grad():
            class_res = class_model(batch_images)
            class_loss = torch.nn.functional.cross_entropy(
                class_res, batch_class_labels, reduction="none"
            )
            class_softmax = torch.nn.functional.softmax(class_res, dim=1)
            class_pred_labels = torch.argmax(class_res, dim=1)
            class_pred_success = class_pred_labels == batch_class_labels

            for i in range(len(batch_class_labels)):
                object_class = class_available[batch_class_labels[i].item()]
                class_results["loss"][object_class] += class_loss[i].item()
                class_results["pred_loss"][object_class] += class_softmax[i][
                    batch_class_labels[i]
                ].item()
                # Start fix (Riccardo)
                # class_results["acc"][object_class] += (
                #     class_pred_success[i].item()
                #     * 1.0
                #     / (
                #         (len(theme_available) - 1) * len(seed) * 2
                #     )  # exclude Seed_Images
                # )
                class_results["acc"][object_class] += class_pred_success[i].item()
                # End fix (Riccardo)
                misclassified_as = class_available[class_pred_labels[i].item()]
                class_results["misclassified"][object_class][misclassified_as] += 1
    # Start fix (Riccardo)
    for object_class in class_available:
        total_samples_for_class = sum(class_results["misclassified"][object_class].values())
        if total_samples_for_class > 0:
            class_results["acc"][object_class] = class_results["acc"][object_class] / total_samples_for_class
    # End fix (Riccardo)
    if not dry_run:
        class_output_path = os.path.join(output_dir, f"{cls}_cls.pth")
        torch.save(class_results, class_output_path)
```

What do you think?

---

**Comment by cywinski** [comment]
Hey Riccardo!

Thanks for spotting this issue and sorry for my late reply.

You are absolutely right about the error in `accuracy_unlearncanvas_cls_sweep_fast.py`, I will fix that, thanks!
However, in `accuracy_unlearncanvas_cls_fast.py`, when accuracy is calculated for 
- style unlearning, it is divided by the number of classes https://github.com/cywinski/SAeUron/blob/main/scripts/accuracy_unlearncanvas_cls_fast.py#L166 
- object unlearning, it is divided by the number of themes https://github.com/cywinski/SAeUron/blob/main/scripts/accuracy_unlearncanvas_cls_fast.py#L198

so that seems okay to me. Am I wrong here?

**Comment by cywinski** [comment]
It's because for class unlearning, both themed class image and class image without any theme are considered (see https://github.com/cywinski/SAeUron/blob/81a6c8de65594e15708a06c2d60293bc7f2ac56b/scripts/accuracy_unlearncanvas_cls_fast.py#L113)

**Comment by riccardorenzulli** [comment]
> It's because for class unlearning, both themed class image and class image without any theme are considered (see
> 
> [SAeUron/scripts/accuracy_unlearncanvas_cls_fast.py](https://github.com/cywinski/SAeUron/blob/81a6c8de65594e15708a06c2d60293bc7f2ac56b/scripts/accuracy_unlearncanvas_cls_fast.py#L113)
> 
> Line 113 in [81a6c8d](/cywinski/SAeUron/commit/81a6c8de65594e15708a06c2d60293bc7f2ac56b)
> 
>  for test_theme in theme_available: 
> )

So just to clarify if I understood correctly: in a single input folder for that script there should be 1020 images (50 styles x 20 objects + 20 single-object prompts).
In your case, however, you’re considering 2000 images (50 x 20 x 2), which explains the 2 in the denominator, since you’re duplicating the single-object prompts for some reason (which I didn’t fully catch).

That said, even with the adjusted denominator, the computed metrics might end up overestimated or underestimated, depending on whether those duplicated single-object prompts are classified correctly or not.
I think the most consistent approach is to simply use the 1020 unique images (without counting any duplicates), as I suggested initially, unless I’m missing something 🙂

**Comment by cywinski** [comment]
Yeah, agreed that it's a better approach. Do you want to make a PR, or should I do it?
I'll try in my free time to recalculate the results for the class unlearning for our method

**Comment by riccardorenzulli** [comment]
> Yeah, agreed that it's a better approach. Do you want to make a PR, or should I do it? 

I can make a PR later if you wish

> I'll try in my free time to recalculate the results for the class unlearning for our method

in my previous experiments with the current code and available checkpoints I got
Average UA: 83.65%
Average IRA: 87.87%
Average CRA: 74.14%

with the fix I got
Average UA: 87.16%
Average IRA: 85.57%
Average CRA: 74.14%

PS: Again, amazing work and thanks for the clarifications!

**Comment by cywinski** [comment]
That would be awesome and thank you for this catch, appreciate it!
